aten/src/ATen/native/Pow.cpp

```diff
@@ -6,85 +6,64 @@
+  build_binary_op(maybe_get_output(), base, exp);
 }
 
-Tensor& pow_out(Tensor& result, const Tensor& base, Scalar exp) {
+TORCH_META_FUNC2(pow, Tensor_Scalar) (const Tensor& base, const Scalar exp) {
   // Numpy compatibility check:
   TORCH_CHECK(!(isIntegralType(base.scalar_type(), true) &&
               exp.isIntegral(true) && exp.toLong() < 0),
               "Integers to negative integer powers are not allowed.");
 
   auto common_dtype = at::result_type(base, exp);
-  TORCH_CHECK(at::can_cast(common_dtype, result.scalar_type()),
-           "result type ", common_dtype, " can't be cast to the desired output type ",
-           result.scalar_type());
+  build_unary_op(maybe_get_output(), base.to(common_dtype));
+}
+
+TORCH_META_FUNC2(pow, Scalar) (const Scalar base, const Tensor& exp) {
+    // This overload doesn't directly use TensorIterator. It attempts to short-circuit,
+    // but otherwise redispatches to the Tensor_Tensor overload.
+    auto dtype = at::result_type(base, exp);
+    set_output(0, exp.sizes(), {}, exp.options().dtype(dtype), exp.names());
+}
 
+} // namespace meta
+
+namespace native {
+
+DEFINE_DISPATCH(pow_tensor_tensor_stub);
+DEFINE_DISPATCH(pow_tensor_scalar_stub);
+
+TORCH_IMPL_FUNC(pow_Tensor_Tensor_out) (const Tensor& base, const Tensor& exp, const Tensor& out) {
+  if (exp.dim() == 0 && exp.device().is_cpu() && base.is_cuda()) {
+    at::pow_out(const_cast<Tensor&>(out), base, exp.item()); // redispatch!
+  } else {
+    pow_tensor_tensor_stub(device_type(), *this);
+  }
+}
+
+TORCH_IMPL_FUNC(pow_Tensor_Scalar_out) (const Tensor& base, const Scalar exp, const Tensor& out) {
+  auto common_dtype = at::result_type(base, exp);
   if (exp.equal(0.0)) {
-    resize_output(result, base.sizes());
-    result.fill_(1);
-    namedinference::propagate_names(result, base);
+    out.fill_(1);
   } else if (exp.equal(1.0)) {
-    resize_output(result, base.sizes());
-    result.copy_(base);
-    namedinference::propagate_names(result, base);
+    out.copy_(base);
   } else {
-    auto iter = TensorIterator::unary_op(result, base.to(common_dtype));
-    pow_tensor_scalar_stub(iter.device_type(), iter, exp);
+    pow_tensor_scalar_stub(device_type(), *this, exp);
   }
-  return result;
 }
 
-Tensor& pow_out(Tensor& result, Scalar base, const Tensor& exp) {
+TORCH_IMPL_FUNC(pow_Scalar_out) (const Scalar base, const Tensor& exp, const Tensor& out) {
   if (base.isComplex() && base.toComplexDouble() == 1.0) {
-    resize_output(result, exp.sizes());
-    result.fill_(1);
-    namedinference::propagate_names(result, exp);
+    out.fill_(1);
   } else if (!base.isComplex() && base.toDouble() == 1.0) {
-    resize_output(result, exp.sizes());
-    result.fill_(1);
-    namedinference::propagate_names(result, exp);
+    out.fill_(1);
   } else {
-    native::pow_out(result, c10::scalar_to_tensor(base, exp.device()), exp);
+    at::pow_out(const_cast<Tensor&>(out), c10::scalar_to_tensor(base, exp.device()), exp); // redispatch!
```

Comment:
I think `scalar_to_tensor` here (and not `wrapped_scalar_tensor`) is the reason for #50121, Scalar argument gets incorrect preference for type promotion. @bdhirsh can you please change what's the behavior of your port wrt type pomotion cases in #50121, and if using `wrapped_scalar_tensor` here fixes it.

Reply:
Good catch! I tested that locally and it looks like it fixes the type promotion problem mentioned in that issue, for cpu:
```
>>> torch.pow(2, torch.tensor([17], dtype=torch.uint8), out=torch.tensor([0]))
tensor([0])
>>> torch.pow(2, torch.tensor(17, dtype=torch.uint8), out=torch.tensor(0))
tensor(0)
```

The behavior for cuda looks unchanged, however. In both the 0-dim tensor case and the scalar case, the computation isn't performed with uint8. I just want to confirm- the behavior is still wrong for cuda, correct? That just means that this fix only partially solves the issue.
```
>>> torch.pow(2, torch.tensor([17], dtype=torch.uint8, device='cuda'), out=torch.tensor([0], device='cuda'))
tensor([131072], device='cuda:0')
>>> torch.pow(2, torch.tensor(17, dtype=torch.uint8, device='cuda'), out=torch.tensor(0, device='cuda'))
tensor(131072, device='cuda:0')
```

Reply:
Yeah, the behavior is wrong in cuda.